assets/js/googlesitekit/data/create-fetch-store.test.js

```diff
@@ -0,0 +1,328 @@
+		const { aParam, objParam } = params;
+		return API.get( 'core', 'test', 'some-data', {
+			aParam,
+			objParam,
+		} );
+	},
+	reducerCallback: ( state, response, params ) => {
+		const { aParam } = params;
+		return {
+			...state,
+			data: {
+				...( state.data || {} ),
+				[ aParam ]: response,
+			},
+		};
+	},
+	argsToParams: ( objParam, aParam ) => {
+		invariant( 'object' === typeof objParam, 'objParam is required.' );
+		invariant( 'undefined' !== typeof aParam, 'aParam is required.' );
+		return {
+			objParam,
+			aParam,
+		};
+	},
+};
+
+describe( 'createFetchStore store', () => {
+	let apiFetchSpy;
+	let dispatch;
+	let registry;
+	let select;
+	let storeDefinition;
+	let store;
+
+	beforeAll( () => {
+		API.setUsingCache( false );
+	} );
+
+	beforeEach( () => {
+		registry = createRegistry();
+
+		storeDefinition = createFetchStore( STORE_PARAMS );
+		registry.registerStore( STORE_NAME, storeDefinition );
+		dispatch = registry.dispatch( STORE_NAME );
+		store = registry.stores[ STORE_NAME ].store;
+		select = registry.select( STORE_NAME );
+
+		apiFetchSpy = jest.spyOn( { apiFetch }, 'apiFetch' );
+	} );
+
+	afterAll( () => {
+		API.setUsingCache( true );
+	} );
+
+	afterEach( () => {
+		unsubscribeFromAll( registry );
+		apiFetchSpy.mockRestore();
+	} );
+
+	describe( 'actions', () => {
+		it( 'includes the expected actions', () => {
+			const fetchStoreDefinition = createFetchStore( {
+				baseName: 'SaveSomeData',
+				controlCallback: async () => true,
+				reducerCallback: ( state ) => state,
+			} );
+
+			expect( Object.keys( fetchStoreDefinition.actions ) ).toEqual( [
+				'fetchSaveSomeData',
+				'receiveSaveSomeData',
+			] );
+		} );
+
+		describe( 'fetch', () => {
+			it( 'validates parameters based on argsToParams', () => {
+				const consoleErrorSpy = jest.spyOn( global.console, 'error' );
+
+				dispatch.fetchGetSomeData();
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( 123 );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( {} );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'aParam is required.' );
+
+				consoleErrorSpy.mockClear();
+			} );
+
+			it( 'yields the expected actions for an arguments error', () => {
+				const fetchStoreDefinition = createFetchStore( {
+					baseName: 'SaveSomeData',
+					controlCallback: async () => true,
+					reducerCallback: ( state ) => state,
+					argsToParams: ( requiredParam ) => {
+						invariant( requiredParam, 'requiredParam is required.' );
+						return {
+							requiredParam,
+						};
+					},
+				} );
+
+				const action = fetchStoreDefinition.actions.fetchSaveSomeData();
+
+				// Catch invariant to get exactly the error we expect.
+				let error;
+				try {
+					invariant( false, 'requiredParam is required.' );
+				} catch ( err ) {
+					error = err;
+				}
+
+				expect( action.next().value ).toEqual( {
```

Comment:
For testing generators, I think it might be better to treat the response as a collection rather than checking a specific `yield`. In this case I think we can get all actions with `[ ...action ]` and then we could `expect.arrayContaining` the `{ response, error }`.

Reply:
I tried that first, but we'll want to test with specific values here (receiving `42` / throwing an error below), so I think individual `yield`s are more applicable here.

Reply:
Looking back at this now, I see why you're stepping through the actions manually and that my suggestion indeed would not work for this test due to specific action yields being called or errors thrown directly. However, this test and `yields the expected actions for an error request` feel like they're testing implementation details and directly interact with the generator functions which none of our other datastore tests do.

It seems like the subsequent tests for `makes a network request based on controlCallback` and `dispatches an error if the request fails` cover this same behavior but at the same level of abstraction as all of our other tests, using mocked http requests. 

It could be that these generator tests tests were useful during the initial development but I don't think they're necessary anymore and that the http-based tests give us more "bang for our buck" in terms of coverage, confidence, and consistency.

Thoughts @tofumatt ?

Reply:
Yeah, I'm in agreement with @aaemnnosttv here. Generator tests and testing each yield is testing the implementation probably a bit too much, and it'd be better to test the output than these. Especially as this isn't testing an action yielded until the end.

There are tools like `redux-saga-tester` that allow us to use helpers that let us wait for specific actions to be dispatched if we want to check the flow of actions dispatched inside a generator action—we don't need that here, but if we _do_ want to be testing implementation/yielded action dispatch order of a generator action, we should use tools like that. They're a bit nicer in terms of their API, and also less prone to illegitimate failures; if the order of actions here changed then this test could fail unless the number of `next` calls changed too.